shim/src/functions.js

```diff
@@ -3,83 +3,92 @@
 // https://github.com/google/caja/blob/master/src/com/google/caja/ses/startSES.js
 // https://github.com/google/caja/blob/master/src/com/google/caja/ses/repairES5.js
 
-import { defineProperty, defineProperties, getPrototypeOf, setPrototypeOf } from './commons';
-
 /**
- * The process to repair constructors:
- * 1. Obtain the prototype from an instance of the syntax
- * 2. Create a substitute noop constructor
- * 3. Replace its prototype property with the original prototype
- * 4. Replace its prototype property's constructor with itself
- * 5. Replace its [[Prototype]] slot with the noop constructor of Function
+ * This block replaces the original Function constructor, and the original
+ * %GeneratorFunction% %AsyncFunction% and %AsyncGeneratorFunction%, with
+ * safe replacements that throw if invoked. 
+ *
+ * These are all reachable via syntax, so it isn't sufficient to just 
+ * replace global properties with safe versions. Our main goal is to prevent 
+ * access to the Function constructor through these starting points.
+
+ * After this block is done, the originals must no longer be reachable, unless 
+ * a copy has been made, and funtions can only be created by syntax (using eval) 
+ * or by invoking a previously saved reference to the originals.
  */
 
 // todo: This function is serialized and evaluated outside of the primal
 // realms and it currently can't contain code coverage metrics.
 /* istanbul ignore file */
-function repairFunction(unsafeRec, functionName, functionDecl) {
-  const { unsafeEval, unsafeFunction, unsafeGlobal } = unsafeRec;
+export function repairFunctions() {
```

Comment:
Not that we evaluate the whole shim in the iframe/vm context, it can grab all it needs from the ambient and we don't need to pass foreign intrinsics to it, and take any risk associated with it.